### PR TITLE
homework 1

### PR DESCRIPTION
**Pull request by yuluomeng**
@vm - check out the `format` method in `hw01/src/cs3500/hw01/duration/AbstractDuration.java`

### CODE REVIEW

File: hw01/src/cs3500/hw01/duration/AbstractDuration.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  @Override
+  public Duration plus(Duration that) {
+    return fromSeconds(this.inSeconds() + that.inSeconds());
+  }
+
+  @Override
+  public String format(String template) {
+    String output = "";
```
Comment (vm) [inline hw01/src/cs3500/hw01/duration/AbstractDuration.java]: Look at StringBuilder

File: hw01/src/cs3500/hw01/duration/AbstractDuration.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+  /**
+   * Returns the number of whole minutes in the given number of seconds, less
+   * the number of whole hours.
+   *
+   * @param inSeconds the total number of seconds
+   * @return the number of remaining minutes
+   */
+  protected static int minutesOf(long inSeconds) {
+    return (int) (inSeconds / 60 % 60);
```
Comment (vm) [inline hw01/src/cs3500/hw01/duration/AbstractDuration.java]: Why are you taking a long and returning an int? Is that the spec?

File: hw01/src/cs3500/hw01/duration/AbstractDuration.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      if (c != '%') {
+        output += c;
+      } else {
+        int next_index = i + 1;
+        if (next_index == template.length()) {
+          throw new IllegalArgumentException("malformed template");
+        }
+
+        char next_c = template.charAt(next_index);
+        switch (next_c) {
```
Comment (vm) [inline hw01/src/cs3500/hw01/duration/AbstractDuration.java]: I'd make a new method for this, and return the formatted string, then add it to the String builder. That way you don't need like 10 break statements.

```
stringBuilder.append(formatNextChar(nextChar))
```

File: hw01/src/cs3500/hw01/duration/AbstractDuration.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+  @Override
+  public Duration plus(Duration that) {
+    return fromSeconds(this.inSeconds() + that.inSeconds());
+  }
+
+  @Override
+  public String format(String template) {
+    String output = "";
+    for (int i = 0; i < template.length(); i++) {
```
Comment (vm) [inline hw01/src/cs3500/hw01/duration/AbstractDuration.java]: I think this is clearer than however you would implement this with Stream. But likely you would have to probably take tuples of pairs of elements (aka its gonna look like shit). This works perfectly fine!

**Comment by vm** [comment]
Done with pass.

File: hw01/src/cs3500/hw01/duration/AbstractDuration.java (hunk 3 above)
Comment (vm) [inline hw01/src/cs3500/hw01/duration/AbstractDuration.java]: Btw Java prefers camelCase for everything except constants like `TIMER_LENGTH`

File: hw01/src/cs3500/hw01/duration/AbstractDuration.java (hunk 2 above)
Comment (yuluomeng) [inline hw01/src/cs3500/hw01/duration/AbstractDuration.java]: Not sure - this was included in the code I pulled from the assignment GH.

**Pull request merged**